packages/did/src/index.ts

```diff
@@ -63,27 +63,37 @@ class DID implements IDID {
    * ```
    *
    * @param { Methods } method
+   * @param {string} chain
    * @param {string} id
    *
    * @returns {void}
    */
   // eslint-disable-next-line no-dupe-class-members
-  set(method: Methods | string, id?: string): IDID {
+  set(method: Methods | string, chain?:string, id?: string): IDID {
     if (method.startsWith('did:')) {
       return this._setDid(method);
     }
-    return this._setDid(`did:${method}:${id}`);
+    if (id !== undefined) {
+      return this._setDid(`did:${method}:${chain}:${id}`);
+    }
+    return this._setDid(`did:${method}:${chain}`);
   }
 
   private _setDid(did: string): IDID {
-    const [, method, id] = did.split(':');
-    if (id === undefined) {
+    const [, method, chain, id] = did.split(':');
+    if (chain === undefined) {
       throw new Error('DID must consist of three parts separated by a colon');
     }
-    if (!DID_SCHEME_PATTERNS.NETWORK.test(method)) {
+    if (id === undefined && !DID_SCHEME_PATTERNS.NETWORK.test(method)) {
+      throw new Error('Network must not be empty and consist only of lowcase alphanumerical characters');
+    }
+    if (id !== undefined && !DID_SCHEME_PATTERNS.NETWORK.test(`${method}:${chain}`)) {
       throw new Error('Network must not be empty and consist only of lowcase alphanumerical characters');
     }
-    if (!DID_SCHEME_PATTERNS.ID.test(id)) {
+    if (id !== undefined && !DID_SCHEME_PATTERNS.ID.test(id)) {
+      throw new Error('Id must consist only of alphanumerical characters, dots, minuses and underscores');
+    }
+    if (id === undefined && !DID_SCHEME_PATTERNS.ID.test(chain)) {
```

Comment:
I'm finding this whole block of logic quite confusing. Could you explain it a bit more please?

Maybe to start, I think that `'DID must consist of three parts separated by a colon'` isn't true? We can have a did with 3 parts (`did:ethr:address`) or a DID with 4 parts (`did:ethr:ewc:address`), right?

Maybe this code would be clearer if first the length of the split DID was determined and then the variable names and error messages could be tailored to that?

Reply:
Yeah, this looks confusing to me now when I randomly saw the logic. This definitely needs refactoring and right breakdown.